activerecord/lib/active_record/middleware/database_selector.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require "active_record/middleware/database_selector/resolver"
+
+module ActiveRecord
+  module Middleware
+    # The DatabaseSelector Middleware provides a framework for automatically
+    # swapping from the primary to the replica database connection. Rails
+    # provides a basic framework to determine when to swap and allows for
+    # applications to write custom strategy classes to override the default
+    # behavior.
+    #
+    # The resolver class defines when the application should switch (i.e. read
+    # from the primary if a write occurred less than 2 seconds ago) and an
+    # operations class that sets a value that helps the resolver class decide
+    # when to switch.
+    #
+    # Rails default middleware uses the request's session to set a timestamp
+    # that informs the application when to read from a primary or read from a
+    # replica.
+    #
+    # To use the DatabaseSelector in your application with default settings add
+    # the following options to your environment config:
+    #
+    #   config.active_record.database_selector = { delay: 2.seconds }
+    #   config.active_record.database_resolver = ActiveRecord::Middleware::DatabaseSelector::Resolver
+    #   config.active_record.database_operations = ActiveRecord::Middleware::DatabaseSelector::Resolver::Session
+    #
+    # New applications will include these lines commented out in the production.rb.
+    #
+    # The default behavior can be changed by setting the config options to a
+    # custom class:
+    #
+    #   config.active_record.database_selector = { delay: 2.seconds }
+    #   config.active_record.database_resolver = MyResolver
+    #   config.active_record.database_operations = MyResolver::MySession
+    class DatabaseSelector
+      def initialize(app, resolver_klass = Resolver, operations_klass = Resolver::Session)
+        @app = app
+        @resolver_klass = resolver_klass
+        @operations_klass = operations_klass
+      end
+
+      attr_reader :resolver_klass, :operations_klass
+
+      # Middleware that determines which database connection to use in a mutliple
+      # database application.
+      def call(env)
+        request = ActionDispatch::Request.new(env)
+
+        select_database(request) do
+          @app.call(env)
+        end
+      end
+
+      private
+
+        def select_database(request, &blk)
+          operations = operations_klass.build(request)
+          database_resolver = resolver_klass.call(operations)
+
+          if reading_request?(request)
+            database_resolver.read(&blk)
+          else
+            database_resolver.write(&blk)
+          end
+        end
+
+        def reading_request?(request)
+          request.get? || request.head?
```

Comment:
@eileencodes These methods are private, but is it otherwise a 'public' interface for building database selector subclasses? (I've read the through the docs and PRS. Apologies if iI missed something or if would have been better to ask on one of the ruby on rails lists. Commenting here, even though code has since changed in other ways, because here it where it was introduced.)

My use case: I want to specify that a particular route is *always* connected to the leader, and adjusting `reading_request?` seems the right place to do that.

Reply:
fwiw, on Rails 6.0, this is what I ended up doing

config/application.rb

```ruby
    require Rails.root.join 'lib', 'extensions', 'active_record', 'middleware_database_selector_patch'
    ActiveRecord::Middleware::DatabaseSelector.prepend(MiddlewareDatabaseSelectorPatch)
    if enable_request_db_selection
      config.active_record.database_selector = { delay: 2.seconds }
      config.active_record.database_resolver = ActiveRecord::Middleware::DatabaseSelector::Resolver
      config.active_record.database_resolver_context = ActiveRecord::Middleware::DatabaseSelector::Resolver::Session
    end
```

where

```ruby
module MiddlewareDatabaseSelectorPatch

  def self.prepended(base)
    super
    base.class_eval do
      class_attribute :leader_only_paths, instance_writer: false, default: Set.new
    end
  end

  private

  # https://github.com/rails/rails/blob/6-0-stable/activerecord/lib/active_record/middleware/database_selector.rb
  def reading_request?(request)
    if (is_reading = super)
      is_leader_only_path = leader_only_paths.any? {|leader_only_path|
        request.path&.start_with?(leader_only_path)
      }
      !is_leader_only_path
    else
      is_reading
    end
  end
end
```

so in config/routes I can

```ruby
  ActiveRecord::Middleware::DatabaseSelector.leader_only_paths << "/pghero"
```

I didn't test perf on it

Reply:
We have a similar use case here. We want to send GraphQL queries to our read replica, but GQL queries always go over `POST`. A public API for this would be great.

Reply:
Just saw @tenderlove 's comment that kind of answers this https://github.com/rails/rails/pull/35073#issuecomment-458645257

I'm not sure I understand the `config.ru` part, is there an example floating around anywhere?

Reply:
Thinking through this more, I think a simple API would be to move the `reading_request?(request)` method into the `Resolver` class. 

That way, when you create a custom resolver you have two options:

- Override `reading_request?` to send request to the read/write database based on request (eg. based on request method, path, or headers).
- Override `read_from_primary?` to send queries to the primary/replica database based on whatever you like.

@eileencodes would you accept a PR that adds this?

Reply:
@eileencodes 👋 Thanks for the 👀 I made a topic (and new account) in https://discuss.rubyonrails.org/t/clarify-which-private-methods-of-databaseselector-are-a-public-stable-interface/75934

Reply:
Added a comment too. Thanks @eileencodes for writing this great feature - this question is my only issue with something that's been very useful for us otherwise ❤️